xla/service/gpu/gpu_compiler.cc

```diff
@@ -1549,22 +1549,51 @@ absl::Status GpuCompiler::OptimizeHloPostLayoutAssignment(
   if (options.target_config.has_value()) {
     return *options.target_config;
   }
-  if (!debug_opts.xla_gpu_target_config_filename().empty()) {
+
+  auto target_config_from_filename =
+      [&](std::string filename) -> absl::StatusOr<Compiler::TargetConfig> {
     std::string gpu_target_config_string;
-    TF_RETURN_IF_ERROR(tsl::ReadFileToString(
-        tsl::Env::Default(), debug_opts.xla_gpu_target_config_filename(),
-        &gpu_target_config_string));
+    TF_RETURN_IF_ERROR(tsl::ReadFileToString(tsl::Env::Default(), filename,
+                                             &gpu_target_config_string));
     stream_executor::GpuTargetConfigProto gpu_target_config_proto;
     if (!tsl::protobuf::TextFormat::ParseFromString(gpu_target_config_string,
                                                     &gpu_target_config_proto)) {
       return absl::FailedPreconditionError(
           "Failed to parse GpuTargetConfigProto");
     }
-
     return Compiler::TargetConfig{gpu_target_config_proto};
+  };
+
+  if (!debug_opts.xla_gpu_target_config_filename().empty()) {
+    return target_config_from_filename(
+        debug_opts.xla_gpu_target_config_filename());
   }
+
   if (executor) {
-    return Compiler::TargetConfig{executor};
+    Compiler::TargetConfig target_config = Compiler::TargetConfig{executor};
+    int64_t device_memory_size = target_config.device_description.device_memory_size();
+    if (device_memory_size == -1) {
+      // Checking for device_memory_size == -1 is how we detect that we are running on
+      // Nvidia's software simulator. When running on simulation, discard the config
+      // from StreamExecutor, which can be inaccurate, and load the hard-coded config
+      // from a file.
+      auto cuda_cc = target_config.device_description.cuda_compute_capability();
+      switch (cuda_cc.major) {
+        case se::CudaComputeCapability::AMPERE: {
+          return target_config_from_filename(
+            "xla/tools/hlo_opt/gpu_specs/a100_40.txtpb");
```

Comment:
I'm not sure paths work this way: have you tried it from any folder?

Reply:
Added a comment to clarify. This is assuming that the compiler is invoked from the XLA root directory. If that is not the case, the user can always use --xla_gpu_target_config_filename, which causes the conditional before this one to be taken.

(You may ask, shouldn't we just always use --xla_gpu_target_config_filename then? We could, but it's nice when the compiler can figure out the right thing without the user having to pass flags.)

Reply:
> This is assuming that the compiler is invoked from the XLA root directory

I'm not sure what do you mean? The compiler can be invoked from any directory, and the behavior should not be changing based on that.

Reply:
>  We could, but it's nice when the compiler can figure out the right thing without the user having to pass flags

But then you get a highly-counterintuitive behavior when it differs depending on which folder is the binary called.

Reply:
Alright, let's just use the flag for the simulation case as well, and we can drop this PR.

Reply:
Actually, rather than closing, I'd like to change this PR to detect Nvidia simulation as before, and check-fail if --xla_gpu_target_config_filename is empty, to remind the user to supply the file with the target config. Are you OK w that?

Reply:
Updated the code.